api/desecapi/models/tokens.py

```diff
@@ -130,6 +130,30 @@ def get_policy(self, rrset=None):
             .first()
         )
 
+    def can_write_all_rrsets(self, domain):
+        forbidden = (
+            # Check if token is explicitly prohibited from writing some RRsets in this domain
+            # (priority order 1-4, see /docs/auth/tokens.rst#token-scoping-policies)
+            self.tokendomainpolicy_set.filter(domain=domain)
+            .filter(perm_write=False)
+            .exists()
+            or
+            # Check that the token has no permissive default policy for this domain
+            # (priority order 4) and apply fall-through to domain-independent policies (5-8)
+            (
+                not self.tokendomainpolicy_set.filter(
+                    domain=domain, subname=None, type=None
+                )
+                .filter(perm_write=True)
+                .exists()
+                # fall-through
+                and self.tokendomainpolicy_set.filter(domain=None)
+                .filter(perm_write=False)
+                .exists()
+            )
+        )
+        return not forbidden
```

Comment:
Thought about it - the logic here makes sense to me. Some suggestions for improvement at the bottom of this pamphlet.

Def. A policy with
    - domain, subname, and type not null has priority 1 (P1),
    - domain and subname not null, but type null has priority 2 (P2),
    - domain and type not null, but subname null has priority 3 (P3),
    - domain not null, but subname and type null has priority 4 (P4),
    - ...

Def. A policy is permissive if perm_write=True, otherwise it is restrictive.

Def. A policy matches if all not-null parameters of it match domain and rrset in question.

Def. A policy is applicable if no higher-priority policy matches.

Def. A set of policies is permissive on a given domain and RRset if there is an applicable permissive policy.

Def. A set of policies allows domain deletion if the set of policies is permissive on that domain for all subnames and types.


Def:
- A: There exists a restrictive priority 1-4 policy.
- B: There exists a permissive priority 4 policy.
- C: There exists a restrictive priority 5-8 policy.


The code says:

```
forbidden = (
   A
   or
   (
     not B
     and
     C
   )
)
```

It returns:

```
not forbidden = not (A or (not B and C))
              = (not A and not (not B and C))
              = (not A and (B or not C))
              = (not A and B) or (not A and not C)
```              

`not forbidden` implies that the policies allow domain deletion:


    There is no restrictive priority 1, 2, 3, or 4 policy [not A]
    AND
    There is a permissive priority 4 policy [B]
==> Applicable policy for each subname and type in this domain is either permissive 1, 2, or 3 priority policy or the permissive priority 4 policy.
==> All subname and all types in this domain have permissive applicable policies.
==> Set of policies allows domain deletion.

    There is no restrictive priority 1, 2, 3, or 4 policy [not A]
    AND
    There is no restrictive priority 5, 6, 7, or 8 policy [not C]
==> There is no restrictive policy.
==> All subname and all types in this domain have permissive applicable policies.
==> Set of policies allows domain deletion.

The inverse doesn't hold true - I think there should be a docstring that explains that.

The function name should be something along the lines of "can savely delete" to make clear that there are cases where deletion could be allowed based on the policy set but isn't.

Reply:
done